### ISSUE TITLE
PDE mode installation (automatic)

### ISSUE THREAD

**Issue by jeremydouglass** [gaocegege/Processing.R#59]
In Processing the normal way to use a new mode is:

-  install the standard Processing IDE (PDE)
-  add the mode using `Tools > Add Tool > Modes`.

In practice, this requires:

-  mode.properties
-  a public release RLangMode.zip
-  RLangMode.txt, containing the URL of RLangMode.zip
-  a URL for RLangMode.txt submitted to Processing's http://download.processing.org/contribs

Here it is in more detail:

1. The mode has a `mode.properties` file -- R.Processing currently builds this with ant, but many modes also keep the file checked in to github, e.g.
      -  https://github.com/jdf/processing.py/blob/master/mode/mode.properties
      -  https://github.com/processing/processing-android-archive/blob/master/mode.properties
2. The mode also has a release -- a public zipped up copy of the built mode
3. The mode maintainer also has posted publicly a contributions manager configuration .txt file, often called e.g. `RLangMode.txt`. It contains a listing that includes the URL of the zipped release. The .txt file doesn't have to be part of the mode, but it must be available online. Some modes publish this .txt file via a website, others via github under Releases, for example:
      - http://py.processing.org/PythonMode.txt
      - https://github.com/joelmoniz/REPLmode/releases/download/latest/REPLMode.txt
4. When the mode is ready to include in the PDE Contributions Manager, the mode maintainer (@gaocegege) sends the URL of this .txt file to a Processing core team member (@prisonerjohn).
5. The URL is added to the contrib_generate tool -- the [Modes section of sources.conf](https://github.com/processing/processing-docs/blob/master/contrib_generate/sources.conf#L273) according to the editing rules in the [README.md](https://github.com/processing/processing-docs/blob/master/contrib_generate/README.md).
6. This is used to periodically generate an online "[contribution listing]" -- a live listing of all libraries, modes, example sets etc. that are available for download, checked by all PDE installs. See it here: http://download.processing.org/contribs
7. When the PDE application runs, the Contributions Manager checks the latest contribution listing online and downloads it to e.g. `~/Library/Processing/contribs.txt` -- see: [ContributionListing.java#L44](https://github.com/processing/processing/blob/master/app/src/processing/app/contrib/ContributionListing.java#L44)

      ```
      static final String LISTING_URL = "http://download.processing.org/contribs";
      static final String LOCAL_FILENAME = "contribs.txt";
      ```
8. When a user selects a listing in the Contribution Manager, the latest zipped build of the mode is downloaded, unpacked, and installed in the modes directory, where it becomes available via the mode.properties file.
9. When the developer wants to make a new stable release they:
      -  increment mode.properties
      -  publish a new zipped release
      -  update their contribution manager config txt file, RLangMode.txt, with the new zipped release URL
      -  next time contrib_generate is run the new mode listing becomes available at http://download.processing.org/contribs and can be automatically downloaded into any PDE installation.

**Comment by jeremydouglass** [comment]
At least, this is how I *believe* setup for automatic mode installation works -- I haven't done it before, on either end.

@prisonerjohn might correct us or point us to documentation....

**Comment by jeremydouglass** [comment]
See also #42 manual mode installation -- vs. automatic

**Comment by gaocegege** [comment]
Shiffman said [It's mostly the same process as with libraries and all the info is here:](https://forum.processing.org/two/discussion/comment/102712/#Comment_102712) https://github.com/processing/processing/wiki/Library-Basics#advertising-your-library

**Comment by jeremydouglass** [comment]
Those docs will give you the general idea. Note that `mode.properties`, which we will use -- described above in the issue -- works the same way but is a bit different from `library.properties`.

**Comment by gaocegege** [comment]
I think Processing.R is ready to contribute to the Contribution Manager in PDE.

![image](https://user-images.githubusercontent.com/5100735/28912590-7251134c-7867-11e7-9a69-930c148ecc11.png)

**Comment by gaocegege** [comment]
Now we have http://gaocegege.com/Processing.R/RLangMode.txt and http://gaocegege.com/Processing.R/RLangMode.zip

**Comment by jeremydouglass** [comment]
Very exciting!!

Regarding:

```
version = v1.0.4
```

I think that version needs to be an int...? Browse all examples of version numbers in http://download.processing.org/contribs

**Comment by jeremydouglass** [comment]
```
# A version number that increments once with each release. This is used to 
# compare different versions of the same library, and check if an update is 
# available. You should think of it as a counter, counting the total number of 
# releases you've had.
version = 22  # This must be parsable as an int

# The version as the user will see it. If blank, the version attribute will be 
# used here. This should be a single word, with no spaces.
prettyVersion = 0.98a  # This is treated as a String
```

**Comment by gaocegege** [comment]
Oh yeah, thanks for your review.

**Comment by gaocegege** [comment]
Updated http://gaocegege.com/Processing.R/RLangMode.txt

**Comment by jeremydouglass** [comment]
Nice! Will keeping a counter be done in the deploy script, or done manually?

**Comment by gaocegege** [comment]
Done manually since we won't deploy the new version frequently, IMO.

**Comment by jeremydouglass** [comment]
That's just fine.

IF you want to make it automatic, two script-y ways:

1. Deploy reads the previous version nBer from the existing RLangMode.txt and increments it in the new file.
2. Create a separate file, version.txt, with only the number. Deploy script increments the number in that file, then includes in RLangMode.txt

**Comment by gaocegege** [comment]
Could we contact with Processing members to add the URL to contrib_generate tool?

**Comment by jeremydouglass** [comment]
@prisonerjohn -- the latest Processing.R mode release is ready to be included in the Contribution Manager. Here is the text file:

http://gaocegege.com/Processing.R/RLangMode.txt

Do you add it, or should we file a PR on processing?

**Comment by prisonerjohn** [comment]
Great, no need to PR, I will review it and get back to you as soon as I get the chance.

**Comment by prisonerjohn** [comment]
Hello,

I just looked through the package and have a couple of notes:
1. Please only use a short version string for the "prettyVersion" field. Something like `1.0.4` instead of `Version v1.0.4, commit dc0a317, built 2017/08/03 21:50`. I know that this is what the Python mode is currently doing, but they also will need to change it.
1. Is there a specific reason the mode is named `R Language` and not simply `R`? It doesn't follow convention with the other language modes like `Java` and `Python`.
<img width="130" alt="screen shot 2017-08-04 at 11 02 38 am" src="https://user-images.githubusercontent.com/119702/28974470-860e1720-7904-11e7-9345-a220e3f0e3a4.png">

**Comment by gaocegege** [comment]
@prisonerjohn Thanks for your review. I will update the file soon.

**Comment by gaocegege** [comment]
Hi, I have updated the txt and zip, PTAL :)

http://gaocegege.com/Processing.R/RLangMode.txt

**Comment by jeremydouglass** [comment]
Thank you again, @prisonerjohn -- hopefully this is now ready for the Contribution Manager! Let us know.

**Comment by jeremydouglass** [comment]
@prisonerjohn -- now that your requested changes have been made is there anything that else we should do on our end?

**Comment by prisonerjohn** [comment]
Looks good, thanks! The R Language mode was just added to the Contribution Manager.
<img width="750" alt="screen shot 2017-08-14 at 10 57 40 am" src="https://user-images.githubusercontent.com/119702/29277746-8f63b3e6-80e0-11e7-8496-de0af008a193.png">

**Comment by jeremydouglass** [comment]
Looks great! I just installed Processing.R through the PDE Contributions Manager.

However, the console displayed a warning:

> The version number for “R for Processing” is not a number.  
> Please contact the author to fix it according to the guidelines.

I'm not sure what to make of that. In http://gaocegege.com/Processing.R/RLangMode.txt 

    version = 104

**Comment by gaocegege** [comment]
Yeah, I met the same warning. I refer to the [text in py.processing.org](http://py.processing.org/3/PythonMode.txt)

**Comment by jeremydouglass** [comment]
I wonder if the issue is that the `prettyVersion` should be `1.0.4` -- not `v1.0.4`?  @prisonerjohn do you know what might be causing this warning message on install?

> The version number for “R for Processing” is not a number.
> Please contact the author to fix it according to the guidelines.

**Comment by prisonerjohn** [comment]
@jeremydouglass That might be it, but I'm not sure. I need to take a look at the PDE source and see where the error is logged.

**Comment by jeremydouglass** [comment]
@prisonerjohn looks like it might be in https://github.com/processing/processing/blob/master/app/src/processing/app/contrib/AvailableContribution.java

**Comment by sudheesh001** [comment]
Just ran into the same issue today, I think they do support having it as `1.0.4` as shown here
https://github.com/processing/processing/blob/2fb23f6e02c1e74f154a5615aa66e170c1c2c100/app/src/processing/app/contrib/Contribution.java#L172

Shouldn't changing the string here solve this problem? https://github.com/gaocegege/Processing.R/blob/76afe79e65b7cea4ebe20da21078c22541ca5996/docs/RLangMode.txt#L7 or is the intention to allow processing to support `v1.0.4` as a `prettyValue`?

**Comment by gaocegege** [comment]
I don't think so. REPL mode has `1.2.9b` and it does not show the warning

**Comment by jeremydouglass** [comment]
But the error message is being thrown on `properties.get("version")` -- not prettyVersion.

**Comment by gaocegege** [comment]
I will add some debug logs in Processing today and see what happened when installing the R mode.

**Comment by gaocegege** [comment]
The error is caused by the mode.properties in the zip file, I have fixed it in #218

**Comment by jeremydouglass** [comment]
Great. I see that you posted a new release 105 to github and updated http://gaocegege.com/Processing.R/RLangMode.txt

The Contributions Manager sometimes takes a while (as much as a day or two) to pick up the new changes, but it should roll out automatically in PDE soon!
